### DISCUSSION TITLE
Component in Airflow's architecture the Listener runs in?

### DISCUSSION THREAD

**Discussion by mishrasangeeta87** [apache/airflow#37432] [Q&A] [answered]
I'm developing a listener plugin to monitor both DAG runs and task instances within Apache Airflow. This plugin requires additional custom packages for its functionality. I'm inquiring about the appropriate components within the Airflow architecture where these packages should be installed: the Web Server, Scheduler, or Workers. Any guidance based on the architectural considerations would be greatly appreciated.

**Comment by potiuk** [comment]
All of them. Airlfow expects the same environment everywhere (for now at least). Due to various optimisiations and edge cases, different components can sometimes do different kinds od callbacks/listeners etc. or even do scheduling decisions in unexpected places (for example in workers or dag file preocessors).

So the best answer is: use the same environment for everything.

**Reply by mishrasangeeta87** [reply, marked as answer]
Thank you
